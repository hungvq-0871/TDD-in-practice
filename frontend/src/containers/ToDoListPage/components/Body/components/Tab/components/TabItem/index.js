import React from 'react';
import PropTypes from 'prop-types';
import { NavItem, NavLink } from 'reactstrap';

function TabItem({
  tabId,
  activeTabId,
  content,
  onClick
}) {
  const classnames = active => active ? 'active' : '';

  return (
    <NavItem>
      <NavLink
        className={classnames(tabId === activeTabId)}
        onClick={() => onClick(tabId)}
      >
        {content}
      </NavLink>
    </NavItem>
  )
}

TabItem.propTypes = {
  tabId: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  activeTabId: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired
}

export default TabItem;
