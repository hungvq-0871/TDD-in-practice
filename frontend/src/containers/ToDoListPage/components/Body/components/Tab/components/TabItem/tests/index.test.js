import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import TabItem from '../index';

const renderTabItem = (activeTabId = "1", onClick = () => { }) => {
  return render(<TabItem
    tabId="1"
    activeTabId={activeTabId}
    content="Work"
    onClick={onClick}
  />);
}

describe(`test how "TabItem" is rendered`, () => {
  test(`it should render it's content correctly`, () => {
    const { container } = renderTabItem();
    const navLinkEle = container.querySelector('.nav-link');

    expect(navLinkEle.innerHTML).toBe('Work')
  });

  test(`it's "a" tag should have "active" className when it's "tabTd" equal to a "activeTabId"`, () => {
    const { container } = renderTabItem();
    const navLinkEle = container.querySelector('.nav-link');

    expect(navLinkEle.className.includes('active')).toBe(true);
  });

  test(`it's "a" tag should not have "active" className when it's "tabTd" does not equal to a "activeTabId"`, () => {
    const { container } = renderTabItem("2");
    const navLinkEle = container.querySelector('.nav-link');

    expect(navLinkEle.className.includes('active')).toBe(false);
  });

  test(`it should match it's snapshot`, () => {
    const { container } = renderTabItem();

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe(`test functions around "TabItem"`, () => {
  test(`it's "onclick" function should run on clicked & receive it's "tabTd"`, () => {
    const handleOnClick = jest.fn(tabId => tabId);
    const { container } = renderTabItem('2', handleOnClick);
    const navLinkEle = container.querySelector('.nav-link');

    fireEvent.click(navLinkEle);

    expect(handleOnClick.mock.calls.length).toBe(1);
    expect(handleOnClick.mock.results[0].value).toBe('1');
  });
});
