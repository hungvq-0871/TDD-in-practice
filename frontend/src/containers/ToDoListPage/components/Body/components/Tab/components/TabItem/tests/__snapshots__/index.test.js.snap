// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test how "TabItem" is rendered it should match it's snapshot 1`] = `
<li
  class="nav-item"
>
  <a
    class="active nav-link"
  >
    Work
  </a>
</li>
`;
